File: EnterpriseRag.Api/Controllers/QueryController.cs
using EnterpriseRag.Api.Data;
using EnterpriseRag.Api.Services;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using System;
using System.Linq;
using System.Threading.Tasks;

namespace EnterpriseRag.Api.Controllers
{
    [ApiController]
    [Route("api/v1/query")]
    public class QueryController : ControllerBase
    {
        private readonly IEmbeddingService _embeddingService;
        private readonly MongoDbContext _dbContext;
        private readonly ILlmService _llmService;
        private readonly ILogger<QueryController> _logger;

        public QueryController(
            IEmbeddingService embeddingService,
            MongoDbContext dbContext,
            ILlmService llmService,
            ILogger<QueryController> logger)
        {
            _embeddingService = embeddingService;
            _dbContext = dbContext;
            _llmService = llmService;
            _logger = logger;
        }

        public class SearchRequest
        {
            public string UserQuestion { get; set; } = string.Empty;
            public string? DocumentName { get; set; }
        }

        [HttpPost("search")]
        public async Task<IActionResult> Search([FromBody] SearchRequest request)
        {
            try
            {
                if (string.IsNullOrWhiteSpace(request.UserQuestion))
                {
                    return BadRequest(new { error = "UserQuestion cannot be empty." });
                }

                _logger.LogInformation("Generating embedding for user question.");
                
                // (A) Pass the question to IEmbeddingService to get the float array
                var queryVector = await _embeddingService.GenerateEmbeddingAsync(request.UserQuestion);

                if (queryVector == null || queryVector.Length == 0)
                {
                    return StatusCode(500, new { error = "Failed to generate embedding for the question." });
                }

                _logger.LogInformation("Searching MongoDB for similar chunks.");
                
                // (B) Pass that float array to MongoDbContext.SearchSimilarChunksAsync
                var similarChunks = await _dbContext.SearchSimilarChunksAsync(queryVector, limit: 3, documentName: request.DocumentName);

                // (C) Generate answer and return the payload
                var results = similarChunks.Select(chunk => new
                {
                    sourceFile = chunk.SourceFile,
                    sequenceIndex = chunk.SequenceIndex,
                    content = chunk.Content
                }).ToList();

                _logger.LogInformation("Generating answer using Groq LLM.");
                var contextStrings = similarChunks.Select(c => c.Content).ToList();
                var answer = await _llmService.GenerateAnswerAsync(request.UserQuestion, contextStrings);

                return Ok(new
                {
                    question = request.UserQuestion,
                    answer = answer,
                    sources = results
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "An error occurred during vector search.");
                return StatusCode(500, new
                {
                    error = "An unexpected error occurred during search.",
                    details = ex.Message
                });
            }
        }
    }
}

File: enterprise-rag-node-api/src/controllers/queryController.ts
import { Request, Response } from 'express';
import { generateEmbedding } from '../services/embeddingService';
import { generateAnswer } from '../services/llmService';
import { getDocumentChunksCollection } from '../config/db';

export const search = async (req: Request, res: Response): Promise<void> => {
  try {
    const { userQuestion, documentName } = req.body;
    
    if (!userQuestion) {
      res.status(400).json({ error: "UserQuestion is required." });
      return;
    }

    console.log(`Received question: ${userQuestion} for document: ${documentName}`);

    const queryVector = await generateEmbedding(userQuestion);

    const collection = getDocumentChunksCollection();
    
    const vectorSearchStage: any = {
      $vectorSearch: {
        index: "vector_index",
        path: "Embedding",
        queryVector: queryVector,
        numCandidates: 100,
        limit: 3
      }
    };

    if (documentName) {
      vectorSearchStage.$vectorSearch.filter = { SourceFile: documentName };
    }

    const pipeline = [vectorSearchStage];
    const similarChunks = await collection.aggregate(pipeline).toArray();

    let context = "";
    const results = [];

    for (const chunk of similarChunks) {
      context += `[Document: ${chunk.SourceFile}, Chunk: ${chunk.SequenceIndex}]\n${chunk.Content}\n\n`;
      results.push({
        sourceFile: chunk.SourceFile,
        sequenceIndex: chunk.SequenceIndex,
        content: chunk.Content
      });
    }

    const answer = await generateAnswer(userQuestion, context);

    res.status(200).json({
      answer: answer,
      results: results
    });

  } catch (error: any) {
    console.error("Query error:", error);
    res.status(500).json({ error: "An unexpected error occurred during search.", details: error.message });
  }
};
